packages/react-reconciler/src/ReactInternalTypes.js

```diff
@@ -418,10 +419,10 @@ export type Dispatcher = {
     reducer: ?(S, A) => S,
   ) => [S, (A) => void],
   useFormState?: <S, P>(
-    action: (S, P) => Promise<S>,
-    initialState: S,
+    action: (Awaited<S>, P) => S,
+    initialState: Awaited<S>,
     permalink?: string,
-  ) => [S, (P) => void],
+  ) => [Awaited<S>, (P) => void],
```

Comment:
As far as I can tell, we only need the `Awaited` in `action` and not `initialState` nor return value: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/67266/files

Let me know if I missed some scenarios that should be rejected at the type-level

Reply:
When I tested this [last week](https://github.com/facebook/react/pull/27571#discussion_r1370002010), passing a promise in as initial state, I got this (rather misleading) error:

> async/await is not yet supported in Client Components, only Server Components. This error is often caused by accidentally adding `'use client'` to a module that was originally written for the server.

Reply:
Ah never mind, it [does give you a type error in your variant](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/67266/files#diff-c2659c0150dfd3f977ac080250bcb31a38280181dd2cc5660acb457c55f901b5R148-R149), so it's all fine I guess.